rust/kernel/chrdev.rs

```diff
@@ -20,10 +19,64 @@ use crate::error::{Error, KernelResult};
 use crate::file_operations;
 use crate::types::CStr;
 
+/// Character device.
+///
+/// # Invariants
+///
+/// - [`self.0`] is valid and non-null.
+/// - [`(*self.0).ops`]   is valid and has static lifetime.
+/// - [`(*self.0).owner`] is valid and has static lifetime.
+struct Cdev(*mut bindings::cdev);
+
+impl Cdev {
+    fn alloc(
+        fops: &'static bindings::file_operations,
+        module: &'static crate::ThisModule,
+    ) -> KernelResult<Self> {
+        // SAFETY: FFI call.
+        let cdev = unsafe { bindings::cdev_alloc() };
+        if cdev.is_null() {
+            return Err(Error::ENOMEM);
+        }
+        // SAFETY: `cdev` is valid and non-null since `cdev_alloc()`
+        // returned a valid pointer which was null-checked.
+        unsafe {
+            (*cdev).ops = fops;
+            (*cdev).owner = module.0;
+        }
+        // INVARIANTS: [`self.0`] is valid and non-null.
+        // [`(*self.0).ops`] and [`(*self.0).owner`] are valid and have
+        // static lifetime.
```

Comment:
I would mention why, at least for `owner` since it may confuse a future reader, i.e.:
  - In the case of `ops` it holds because it was coerced from a reference with `'static` lifetime.
  - However, the `owner` case is different: it is because the `module.0` pointer comes from a `ThisModule` (the fact that the reference to the `ThisModule` has static lifetime does not imply the pointer inside `ThisModule` does -- it is the invariants of `ThisModule` that ensure that).

Reply:
I created #211 and #212 to improve `ThisModule` on its invariants.

Reply:
It looks like `owner` has module lifetime, not static lifetime. That means we could potentially cause bad stuff to happen by giving `Cdev::alloc` the `this_module` of a **different** module! I wonder if we need to consider that.

Reply:
(Sorry I haven't followed this closely, I'll take a closer look tomorrow. For now just answer the question above.)

We need to consider it. The ultimate one should ask to help answer this is: can someone write code without `unsafe` blocks that violate memory safety? If the answer is yes, then we need to do something about it; in general we have two possible routes: mark the function `unsafe` and document the safety requirements that callers are required to manually satisfy, or (preferred) find a way to prevent abuse of the function (e.g., by not allowing users to specify arbitrary modules) though it's not always possible...

Reply:
Yes, `ThisModule` is only to be used only within the current module (the constructor is unsafe to ensure proper usage, but I opened those two issues because it should be improved). In fact, I think nobody should be creating any `ThisModule` object in normal circumstances, only using the one provided via the `module!` macro (which is `static`). We can create a lint to ensure this later on.

Reply:
Shall I put a `TODO` in the `SAFETY` comments for now re. `owner`?

Reply:
We should ensure nobody falls into that trap, so I wouldn't put in `chrdev.rs` but in `ThisModule`. Probably we should create a `get()` unsafe method in `ThisModule` and explain that you cannot pass it to another module. I will add this point to the issue I created above.

Reply:
I just noticed that we might be in a serious spot of lifetime trouble here. I'll push the updated `SAFETY`s, and place a comment outlining the problem.